cmd/trace-agent/sampler.go

```diff
@@ -75,46 +80,55 @@ func (s *Sampler) Add(t agent.ProcessedTrace) (sampled bool, rate float64) {
 
 // Stop stops the sampler
 func (s *Sampler) Stop() {
+	close(s.exit)
```

Comment:
If we want to be 100% synchronized here, I'd say:
```go
s.exit <- struct{}{}
<-s.exit
```
This way we send the signal to exit the loop and only continue after that goroutine has exited. To get the signal from the goroutine we would put:
```go
defer close(s.exit)
```
...at the top of `logStats()`

Reply:
Are you sure you need blocking behaviour here? I see that some other Stops are asynchronous (eg.sampler/ presampler.go). OTOH stats/concentrator.go uses WaitGroup to synchronize stop (which I think is a cleaner way than logstats closing the channel since closing twice panics, and there may be more than one routine you would want to stop in general).

Also, is the following comment stale?
https://github.com/DataDog/datadog-agent/blob/04bfbfced8e9342d01ecf61a41b07c2999778ddd/pkg/trace/agent/agent.go#L122

Reply:
> I see that some other Stops are asynchronous (eg.sampler/ presampler.go)

We are in the process of improving the trace-agent. That behavior is not ideal and I wouldn't use it as an example. It's not what I'd call a clean exit 😅

> OTOH stats/concentrator.go uses WaitGroup to synchronize stop

The same as above applies here. I consider a "wait group" to be for _groups_, not for a single goroutine. It makes reading the code confusing. The concentrator has an exit channel _and_ an exit wait group which confuses things even further.

> Also, is the following comment stale?

I think what is said in that comment would be a nice change, but I consider it just an aesthetical cleanup that isn't of any priority at the moment; I'd disregard it.

> there may be more than one routine

Then we would use a wait group. If there is only one routine, I would do what I've suggested, but I'm open to alternatives if you know any better way.

> closing twice panics

Is this the case in our situation? If not, should we worry about it?

I don't feel strongly about my suggestions and am open to changes. I'm just thinking that if we are to synchronize things, we might as well do it right the sense that if we send a shutdown signal we should wait and ensure the thing actually got shut down. A `sync.WaitGroup` in my opinion makes sense when there is an actual _group_, otherwise it can be confusing to a reader.

Reply:
Also, for the record, we've had some very sensitive situations with working towards exiting goroutines cleanly in [dd-trace-go](https://github.com/DataDog/dd-trace-go) and this method seemed to work very efficiently.